concurrency/parallel_worker.go

```diff
@@ -0,0 +1,47 @@
+package concurrency
+
+import "sync"
+
+// ParallelWorker is a struct providing an helper to run tasks in parallel
+// while capping the parallelism with a known number of workers.
+type ParallelWorker struct {
+	sem         chan struct{}
+	wg          *sync.WaitGroup
+	endFunction func()
+}
+
+// NewParallelWorker constructs a new parallel worker running at maximum
+// <workers> jobs at a time. <endFunc> is a callback called when all the jobs
+// are over and that "Stop" is called.
+func NewParallelWorker(workers int, endFunc func()) ParallelWorker {
+	w := ParallelWorker{
+		sem:         make(chan struct{}, workers),
+		wg:          &sync.WaitGroup{},
+		endFunction: endFunc,
+	}
+
+	return w
+}
+
+// Stop should be called to release all resources. It waits for all the runnings
+// tasks to be over Calling Perform after this would result in a panic
+func (w ParallelWorker) Stop() {
```

Comment:
I don't find the function name good. I think it does not reflect the fact that it's actually a wait. Maybe `WaitAndDone` or `WaitAndFinish`?

Reply:
ACK

Reply:
@EtienneM would `Complete()` would be better? =p

Reply:
I'm not sure. The `Stop` method is really just a wait operation. And from my perspective, `Complete` is not clear about the face that the method will block until the computation is over.

Reply:
I've used `CompleteProcessing`

In English it sounds semantically quite good to me, what do you think?

Reply:
Yep OK, thanks :)